applications/asset_tracker_v2/src/cloud/azure_iot_hub_integration.c

```diff
@@ -0,0 +1,329 @@
+		LOG_DBG("AZURE_IOT_HUB_EVT_DPS_CONNECTING");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_REGISTERING:
+		LOG_DBG("AZURE_IOT_HUB_EVT_DPS_REGISTERING");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_DONE:
+		LOG_DBG("AZURE_IOT_HUB_EVT_DPS_DONE");
+		break;
+	case AZURE_IOT_HUB_EVT_DPS_FAILED:
+		LOG_DBG("AZURE_IOT_HUB_EVT_DPS_FAILED");
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RECEIVED:
+		LOG_DBG("AZURE_IOT_HUB_EVT_TWIN_RECEIVED");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_DATA_RECEIVED;
+		cloud_wrap_evt.data.buf = evt->data.msg.ptr;
+		cloud_wrap_evt.data.len = evt->data.msg.len;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED:
+		LOG_DBG("AZURE_IOT_HUB_EVT_TWIN_DESIRED_RECEIVED");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_DATA_RECEIVED;
+		cloud_wrap_evt.data.buf = evt->data.msg.ptr;
+		cloud_wrap_evt.data.len = evt->data.msg.len;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_DIRECT_METHOD:
+		LOG_DBG("AZURE_IOT_HUB_EVT_DIRECT_METHOD");
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS:
+		LOG_DBG("AZURE_IOT_HUB_EVT_TWIN_RESULT_SUCCESS");
+		break;
+	case AZURE_IOT_HUB_EVT_TWIN_RESULT_FAIL:
+		LOG_DBG("AZURE_IOT_HUB_EVT_TWIN_RESULT_FAIL");
+		break;
+	case AZURE_IOT_HUB_EVT_PUBACK:
+		LOG_DBG("AZURE_IOT_HUB_EVT_PUBACK");
+		break;
+	case AZURE_IOT_HUB_EVT_ERROR:
+		LOG_DBG("AZURE_IOT_HUB_EVT_ERROR");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_ERROR;
+		cloud_wrap_evt.err = evt->data.err;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_FOTA_START:
+		LOG_DBG("AZURE_IOT_HUB_EVT_FOTA_START");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_FOTA_START;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_FOTA_DONE:
+		LOG_DBG("AZURE_IOT_HUB_EVT_FOTA_DONE");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_FOTA_DONE;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_FOTA_ERASE_PENDING:
+		LOG_DBG("AZURE_IOT_HUB_EVT_FOTA_ERASE_PENDING");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_FOTA_ERASE_PENDING;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_FOTA_ERASE_DONE:
+		LOG_DBG("AZURE_IOT_HUB_EVT_FOTA_ERASE_DONE");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_FOTA_ERASE_DONE;
+		notify = true;
+		break;
+	case AZURE_IOT_HUB_EVT_FOTA_ERROR:
+		LOG_DBG("AZURE_IOT_HUB_EVT_FOTA_ERROR");
+		cloud_wrap_evt.type = CLOUD_WRAP_EVT_FOTA_ERROR;
+		notify = true;
+		break;
+	default:
+		LOG_ERR("Unknown Azure IoT Hub event type: %d", evt->type);
+		break;
+	}
+
+	if (notify) {
+		cloud_wrapper_notify_event(&cloud_wrap_evt);
+	}
+}
+
+int cloud_wrap_init(cloud_wrap_evt_handler_t event_handler)
+{
+	int err;
+
+#if !defined(CONFIG_CLOUD_CLIENT_ID_USE_CUSTOM)
+	char imei_buf[20];
+
+	/* Retrieve device IMEI from modem. */
+	err = at_cmd_write("AT+CGSN", imei_buf, sizeof(imei_buf), NULL);
+	if (err) {
+		LOG_ERR("Not able to retrieve device IMEI from modem");
+		return err;
+	}
+
+	/* Set null character at the end of the device IMEI. */
+	imei_buf[AZURE_IOT_HUB_CLIENT_ID_LEN] = 0;
+
+	strncpy(client_id_buf, imei_buf, sizeof(client_id_buf) - 1);
```

Comment:
I think I'd use `snprintk` instead, as `strncpy` doesn't null-terminate and we should ensure the client ID is indeed null-terminated